api/shared/wrap_for_cf_on_k8s.go

```diff
@@ -0,0 +1,124 @@
+package shared
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/client-go/transport"
+
+	"code.cloudfoundry.org/cli/actor/v7action"
+	"code.cloudfoundry.org/cli/command"
+
+	// imported for the side effects
+	_ "k8s.io/client-go/plugin/pkg/client/auth/azure"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/oidc"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 net/http.RoundTripper
+
+func WrapForCFOnK8sAuth(config command.Config, k8sConfigGetter v7action.KubernetesConfigGetter, roundTripper http.RoundTripper) (http.RoundTripper, error) {
+	username, err := config.CurrentUserName()
+	if err != nil {
+		return nil, err
+	}
+	if username == "" {
+		return nil, errors.New("current user not set")
+	}
+
+	k8sConfig, err := k8sConfigGetter.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	restConfig, err := clientcmd.NewDefaultClientConfig(
+		*k8sConfig,
+		&clientcmd.ConfigOverrides{
+			Context: api.Context{AuthInfo: username},
+		},
+	).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Special case for certs, since we don't want mtls
+	cert, err := getCert(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	transportConfig, err := restConfig.TransportConfig()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get transport config: %w", err)
+	}
+
+	if cert != nil {
+		return certRoundTripper{
+			cert:         cert,
+			roundTripper: roundTripper,
+		}, nil
+	}
+
+	if transportConfig.WrapTransport == nil {
+		// i.e. not auth-provider or exec plugin
+		return transport.HTTPWrappersForConfig(transportConfig, roundTripper)
+	}
+
+	// using auth provider to generate token
+	return transportConfig.WrapTransport(roundTripper), nil
```

Comment:
We have recently found our that [this check is redundant](https://github.com/kubernetes/client-go/blob/6889a6b513d4b6b374bec632433d39f5aede033d/transport/round_trippers.go#L39-L42).

Reply:
should that check be removed before we merge the PR?

Reply:
Yes, we might get away with just `return transportConfig.HTTPWrappersForConfig(...)` but I'd try it first, just in case.